### PR TITLE
Pauli Z measurement

### PR DESCRIPTION
**Pull request by juliafreund**

### CODE REVIEW

**Comment by jwallnoefer** [comment]
* needs doc-strings and other measurements as discussed
* also needs test

File: src/graphepp/graphepp.py [hunk 1]
```diff
@@ -52,6 +52,15 @@ def adj_matrix(N, E):
     return adj
 
 
+def local_PauliZ(graph, n):
+    edges = list(graph.E)
+    # remove all edges which contain node n
+    newEdges = tuple((x, y) for (x, y) in edges if x != n)
+    newEdges = tuple((x, y) for (x, y) in newEdges if y != n)
```
Comment (con-f-use) [inline src/graphepp/graphepp.py]:
```suggestion
    newEdges = tuple((x, y) for x, y in edges if x != n and y != n)
```

Comment (juliafreund) [reply src/graphepp/graphepp.py]: well, well good point :)

File: src/graphepp/graphepp.py [hunk 2]
```diff
@@ -52,6 +52,92 @@ def adj_matrix(N, E):
...
+
+    """
+    # find all neighbors of qubit n
+    neighbors = np.array([(x, y) for (x, y) in list(graph.E) if x == n or y == n])
+
+    # if the vertex has neighbors, we perform the measurement, else we return the same graph state
+    if neighbors.shape[0] != 0:
+        # if neighbor is not specified, choose the first neighbor in the array
+        if neighbor == -1:
+            neighbor = neighbors[0][0] if neighbors[0][0] != n else neighbors[0][1]
```
Comment (jwallnoefer) [inline src/graphepp/graphepp.py]: This seems like a lot of unnecessary index shuffling. I would also add a protection against specifying an invalid neighbor.
```suggestion
    neighbors = np.nonzero(graph.adj[n, :]
    if neighbor == -1:
        neighbor = neighbors[0]
    else:
        if neighbor not in neighbors:
            raise ValueError(f"neighbor={neighbor} is not a neighbor of vertex n={n}.")
```

File: src/graphepp/graphepp.py [hunk 3]
```diff
@@ -52,6 +52,92 @@ def adj_matrix(N, E):
...
+    ----------
+    graph : Graph
+        Instance of Graph class
+    n : int
+        index of the qubit on which the local Pauli Y measurement is performed
+    neighbor : int
+        neighboring qubit on which the local complementation is performed, optional parameter
+    Returns
+    -------
+    Graph : Graph
```
Comment (jwallnoefer) [inline src/graphepp/graphepp.py]:
```suggestion
    Graph
```

File: src/graphepp/graphepp.py [hunk 4]
```diff
@@ -52,6 +52,92 @@ def adj_matrix(N, E):
...
+    Parameters
+    ----------
+    graph : Graph
+        Instance of Graph class
+    n : int
+        index of the qubit on which the local Pauli Y measurement is performed
+
+    Returns
+    -------
+    Graph : Graph
```
Comment (jwallnoefer) [inline src/graphepp/graphepp.py]:
```suggestion
    Graph
```

File: src/graphepp/graphepp.py [hunk 5]
```diff
@@ -52,6 +52,92 @@ def adj_matrix(N, E):
...
+    Parameters
+    ----------
+    graph : Graph
+        Instance of Graph class
+    n : int
+        index of the qubit on which the local Pauli Z measurement is performed
+
+    Returns
+    -------
+    Graph : Graph
```
Comment (jwallnoefer) [inline src/graphepp/graphepp.py]:
```suggestion
    Graph
```

File: tests/measurements_test.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+""" Unittest File for verification of the Pauli measurements
+
+"""
+
+
+import pytest
+import numpy as np
+from unittest.mock import MagicMock
```
Comment (jwallnoefer) [inline tests/measurements_test.py]: This is not used, so delete it?
```suggestion
```

File: tests/measurements_test.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+def test_PauliY():
+    test_state = gg.Graph(3, ((0, 1), (1, 0), (1, 2), (2, 0)))
+    test_state_expectation = gg.Graph(3, ((),))
+    test_state_measured = gg.local_PauliY(test_state, n=1)
+
+    assert test_state_measured == test_state_expectation
+
+
+def test_PauliX():
```
Comment (jwallnoefer) [inline tests/measurements_test.py:25]: Maybe also add a test that the default case of specifying no neighbor is handled correctly?

**Review by jwallnoefer — COMMENTED** [review]
Maybe we could also rename the new functions? maybe measure_Z etc. to make it explicit that the graph transformations are related to measurements?

I would also suggest moving the functions to the "auxiliary functions" part of the file and not at the very top of the file (tbh I should probably split this up in multiple files at some point)

File: src/graphepp/graphepp.py [hunk 8]
```diff
@@ -651,6 +651,92 @@ def complement_state(rho, n, graph):
...
+    -------
+    Graph
+        Returns a new Graph state with updated edge set
+
+    """
+
+    # find all neighbors of qubit n
+    neighbors = np.nonzero(graph.adj[n, :])[0]
+
+    print(neighbors)
```
Comment (jwallnoefer) [inline src/graphepp/graphepp.py]:
```suggestion
```

**Pull request merged**